docs/RingBuffer.md

```diff
@@ -0,0 +1,414 @@
+};
+
+#define ring_buffer_opts__last_field sz
+
+/**
+ * @brief Creates a new ring buffer manager.
+ *
+ * @param[in] map_fd File descriptor to ring buffer map.
+ * @param[in] sample_cb Pointer to ring buffer notification callback function.
+ * @param[in] ctx Pointer to sample_cb callback function context.
+ * @param[in] opts Ring buffer options.
+ *
+ * @returns Pointer to ring buffer manager.
+ */
+struct ring_buffer *
+ring_buffer__new(int map_fd, ring_buffer_sample_fn sample_cb, void *ctx,
+		 const struct ring_buffer_opts *opts);
+
+/**
+ * @brief poll ringbuf for new data
+ * Poll for available data and consume records, if any are available.
+ * Returns number of records consumed (or INT_MAX, whichever is less), or
+ * negative number, if any of the registered callbacks returned error.
+ *
+ * If timeout_ms is zero, poll will not wait but only invoke the callback on records that are ready.
+ * If timeout_ms is -1, poll will wait until data is ready (no timeout).
+ *
+ * This function is only supported when the RINGBUF_FLAG_NO_AUTO_CALLBACK flag is set.
+ *
+ * @param[in] rb Pointer to ring buffer manager.
+ * @param[in] timeout_ms maximum time to wait for (in milliseconds).
+ *
+ * @returns number of records consumed, INT_MAX, or a negative number on error
+ */
+int ring_buffer__poll(struct ring_buffer *rb, int timeout_ms);
+
+/**
+ * @brief Frees a ring buffer manager.
+ *
+ * @param[in] rb Pointer to ring buffer manager to be freed.
+ */
+void ring_buffer__free(struct ring_buffer *rb);
+```
+
+### New ebpf APIs for mapped memory consumer
+
+```c
+/**
+ * get pointers to mapped producer and consumer pages
+ *
+ * @param[in] map_fd File descriptor to ring buffer map.
+ * @param[out] producer pointer* to start of read-only mapped producer pages
+ * @param[out] consumer pointer* to start of read-write mapped consumer page
+ *
+ * @returns EBPF_SUCCESS on success, or error
+ */
+ebpf_result_t ebpf_ring_buffer_get_buffer(fd_t map_fd, void **producer, void **consumer);
+
+/**
+ * get the wait handle to use with WaitForSingleObject/WaitForMultipleObject
+ *
+ * @param[in] map_fd File descriptor to ring buffer map.
+ *
+ * @returns Wait handle
+ */
+HANDLE ebpf_ring_buffer_get_wait_handle(fd_t map_fd);
+```
+
+### New user-space helpers for memory mapped consumer
+
+```c
+/**
+ * The below helpers simplify memory-mapped consumer logic
+ * by abstracting operations on the producer and consumer offsets.
+ */
+
+/**
+ * Get pointer to consumer offset from consumer page.
+ *
+ * @param[in] cons pointer* to start of read-write mapped consumer page
+ *
+ * @returns Pointer to consumer offset
+ */
+uint64_t* rb__consumer_offset(void *cons);
+
+/**
+ * Get pointer to producer offset from producer page.
+ *
+ * @param[in] prod pointer* to start of read-only mapped producer pages
+ *
+ * @returns Pointer to producer offset
+ */
+volatile const uint64_t* rb__producer_offset(volatile const void *prod);
```

Comment:
Does this exact same prototype exist on Linux? I couldn't find it, I could only find
https://github.com/torvalds/linux/blob/master/include/uapi/linux/bpf.h#L4558